src/core/io_queue.cc

```diff
@@ -683,7 +683,16 @@ io_group::io_group(io_queue::config io_cfg, unsigned nr_queues)
         const auto& fg = _fgs[g_idx];
         auto max_cap = fg.maximum_capacity();
         for (unsigned shift = 0; ; shift++) {
-            auto tokens = internal::request_tokens(io_direction_and_length(idx, 1 << (shift + io_queue::block_size_shift)), _config);
+            // Break this loop if the calculated request length gets larger than the
+            // request length limit. This avoids an infinite loop if the
+            // configured io_properties are huge and capacity ends up being always 0.
+            // _max_request_length will hold default values of io_group::request_length_limit
+            unsigned long req_length = 1ul << (shift + io_queue::block_size_shift);
+            if (req_length > io_group::request_length_limit) {
```

Comment:
Thinking out loud.

Let's assume we hit that if, so req_length > request_length_limit. Then we set max request size to be request_length_limit value, but we don't know what's `request_tokens(request_length_limit) <=> max_cap` -- it can happen (in theory) that `request_tokens(request_length_limit) > max_cap` which is not correct.

Reply:
@xemul you mean we end up with tokens(max(2^i)) < tokens(request_length_limit) < max_cap and we choose max(2^i) ?

Reply:
Yes

Reply:
I agree, but I'm ok with this, the current algorithm is "choose the biggest power of two which results in capacity smaller than max_cap", that behavior is preserved if we move the cut off point below. 
let me see if I can cook something to fine tune this.

Reply:
I added a short snippet to account for that. The path shouldn't be too hot IIUC to worry about the extra calculations.